### PR TITLE
Docker container

### PR DESCRIPTION
**Pull request by aprokop**
- Updating clang to 8.0 will require updating clang-format
- Not tested yet as there is no build system in place
- Unclear what version of OpenMPI to use
  Would like to match to Spectrum MPI installed on Summit

### CODE REVIEW

File: docker/.gitignore [hunk 1]
```diff
@@ -0,0 +1,3 @@
+docker-compose.override.yml
+Dockerfile_*_generated
```
Comment (dalg24) [inline docker/.gitignore]:
```suggestion
```

File: docker/Dockerfile_stack [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+
+RUN apt-get update && apt-get install -y \
+        build-essential \
+        bc \
+        gfortran \
+        curl \
+        git \
+        wget \
+        vim \
+        emacs \
```
Comment (dalg24) [inline docker/Dockerfile_stack]:
```suggestion
```

File: docker/Dockerfile_stack [hunk 3]
```diff
@@ -0,0 +1,219 @@
+ARG BASE=ubuntu:16.04
+FROM $BASE
+
+ARG NPROC=4
+
+RUN apt-get update && apt-get install -y \
+        build-essential \
+        bc \
+        gfortran \
```
Comment (dalg24) [inline docker/Dockerfile_stack]:
```suggestion
```

File: docker/Dockerfile_stack [hunk 4]
```diff
@@ -0,0 +1,219 @@
+ARG BASE=ubuntu:16.04
```
Comment (dalg24) [inline docker/Dockerfile_stack]:
```suggestion
ARG BASE=ubuntu:18.04
```

File: docker/Dockerfile_stack [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+        zlib1g-dev \
+        python2.7-dev \
+        ninja-build \
+        doxygen \
+        && \
+    apt-get install -y software-properties-common && \
+    add-apt-repository -y ppa:ubuntu-toolchain-r/test && \
+    apt-get update && apt-get install -y \
+        gcc-7 \
+        g++-7 && \
```
Comment (dalg24) [inline docker/Dockerfile_stack]: gcc-7 is the default on Ubuntu18.04

File: docker/Dockerfile_stack [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+    ${OPENMPI_SOURCE_DIR}/configure --prefix=${OPENMPI_INSTALL_DIR} && \
+    make -j${NPROC} install && \
+    rm -rf ${OPENMPI_ARCHIVE} && \
+    rm -rf ${OPENMPI_BUILD_DIR} && \
+    rm -rf ${OPENMPI_SOURCE_DIR}
+
+ENV MPI_DIR=/opt/openmpi/3.1
+# Put OPENMPI_DIR at the end of the path so that /ust/local/bin/mpiexec will
+# overwrite it
+ENV PATH=$PATH:${MPI_DIR}/bin
```
Comment (dalg24) [inline docker/Dockerfile_stack]: Prepend it

File: docker/Dockerfile_stack [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+    export OPENMPI_SOURCE_DIR=${PREFIX}/source/openmpi/${OPENMPI_VERSION} && \
+    export OPENMPI_BUILD_DIR=${PREFIX}/build/openmpi/${OPENMPI_VERSION} && \
+    export OPENMPI_INSTALL_DIR=/opt/openmpi/${OPENMPI_VERSION_SHORT} && \
+    wget --quiet ${OPENMPI_URL} --output-document=${OPENMPI_ARCHIVE} && \
+    echo "${OPENMPI_SHA1} ${OPENMPI_ARCHIVE}" | sha1sum -c && \
+    mkdir -p ${OPENMPI_SOURCE_DIR} && \
+    tar -xf ${OPENMPI_ARCHIVE} -C ${OPENMPI_SOURCE_DIR} --strip-components=1 && \
+    mkdir -p ${OPENMPI_BUILD_DIR} && \
+    cd ${OPENMPI_BUILD_DIR} && \
+    ${OPENMPI_SOURCE_DIR}/configure --prefix=${OPENMPI_INSTALL_DIR} && \
```
Comment (dalg24) [inline docker/Dockerfile_stack]: Don't we need a flag for CUDA-aware MPI?

File: docker/Dockerfile_stack [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+    rm -rf ${KOKKOS_BUILD_DIR} && \
+    rm -rf ${KOKKOS_SOURCE_DIR} && \
+    rm /root/kokkos.patch" ]
+
+
+ENV KOKKOS_DIR=/scratch/source/kokkos-tools/release/src/tools
+
+# install Kokkos Profiling and Debugging Tools
+# sh does not support arrays so we need to use bash
+# Note: the commit hash provided below has been tagged as Version 2.5.00 on GitHub
```
Comment (dalg24) [inline docker/Dockerfile_stack]: Why did you not match the Kokkos version?

File: docker/Dockerfile_stack [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+    mkdir build && cd build && \
+    cmake -D CMAKE_BUILD_TYPE=Release -D CMAKE_INSTALL_PREFIX=${BENCHMARK_DIR} .. && \
+    make -j${NPROC} && make install && \
+    cd ../.. && rm -rf benchmark
+
+# append the option flag --allow-run-as-root to mpiexec
+RUN echo '#!/usr/bin/env bash' > /usr/local/bin/mpiexec && \
+    echo '${MPI_DIR}/bin/mpiexec --allow-run-as-root "$@"' >> /usr/local/bin/mpiexec && \
+    chmod +x /usr/local/bin/mpiexec && \
+    mpiexec --version
```
Comment (dalg24) [inline docker/Dockerfile_stack]: Remove that block

**Comment by aprokop** [comment]
Chatted with some OLCF folks:
- There is no easy way to find with version of OpenMPI IBM's Spetrum-MPI matches
- From some chats with IBM right now it seems to be version 3
- There may be some plans of IBM to move to 4 later in the year
- But there is not guarantee that increment in major version of Spectrum-MPI mean incrementing major version of OpenMPI

**Comment by aprokop** [comment]
@dalg24 Addressed your comments. So far, tested non-cuda container only. DistrubutedSearchTree test segfaults, have not investigated why yet. Things to do:
- Decide on how many containers we want to build
  DTK builds 3, Cabana 1, I think at least two (cuda and non-cuda)
- How to organize files
  Where to put build script, how to name directories, etc.

**Comment by dalg24** [comment]
Try to revert cf431e2d084b8cd9318136f228f3ddd2652d8e5e

**Comment by dalg24** [comment]
retest this please

**Comment by dalg24** [comment]
retest this please

**Comment by dalg24** [comment]
retest this please

**Comment by dalg24** [comment]
retest this please

**Comment by dalg24** [comment]
retest this please

**Comment by dalg24** [comment]
NOT ready.  Just trying to figure CI out...

**Comment by dalg24** [comment]
test this please

**Comment by dalg24** [comment]
retest this please

**Comment by dalg24** [comment]
retest this please

File: docker/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+ENV PATH=${LLVM_DIR}/bin:$PATH
+
+# Install OpenMPI
+ENV OPENMPI_DIR=/opt/openmpi
+RUN OPENMPI_VERSION=3.1.3 && \
+    OPENMPI_VERSION_SHORT=3.1 && \
+    OPENMPI_SHA1=b3c60e2bdd5a8a8e758fd741f9a5bebb84da5e81 && \
+    OPENMPI_URL=https://download.open-mpi.org/release/open-mpi/v${OPENMPI_VERSION_SHORT}/openmpi-${OPENMPI_VERSION}.tar.bz2 && \
+    OPENMPI_ARCHIVE=openmpi-${OPENMPI_VERSION}.tar.bz2 && \
+    [ ! -z "${CUDA_VERSION}" ] && CUDA_OPTIONS=--with-cuda || true && \
```
[the remark this answers is not in the document]
Comment (dalg24) [reply docker/Dockerfile]: Defined in `nvidia/cuda`

File: docker/jenkins/build.sh [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+cd $ARBORX_DIR
+rm -rf build
+mkdir build
+cd build
+
+cmake "${ARGS[@]}" "${EXTRA_ARGS[@]}" "${ARBORX_DIR}"
+
+make -j${NPROCS}
+
+ctest --no-compress-output -T Test
```
[the remark this answers is not in the document]
Comment (dalg24) [reply docker/jenkins/build.sh]: Yes we have not set the number of procs/threads required by each test.

File: docker/Dockerfile [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+    wget --quiet ${KOKKOS_URL} --output-document=${KOKKOS_ARCHIVE} && \
+    mkdir -p kokkos && \
+    tar -xf ${KOKKOS_ARCHIVE} -C kokkos --strip-components=1 && \
+    cd kokkos && \
+    patch -p1 < /root/kokkos.patch && \
+    mkdir -p build && cd build && \
+    [ ! -z "${CUDA_VERSION}" ] && CUDA_OPTIONS="--with-cuda --with-cuda-options=enable_lambda,enable_uvm" || true && \
+    ../generate_makefile.bash --prefix=${KOKKOS_DIR} \
+       --with-serial --with-openmp ${CUDA_OPTIONS} \
+       --arch=SNB,Volta70 \
```
[the remark this answers is not in the document]
Comment (dalg24) [reply docker/Dockerfile]: Will make it a build argument in a later PR

File: docker/Dockerfile [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+    mkdir -p ${CMAKE_DIR} && \
+    sh ${CMAKE_SCRIPT} --skip-license --prefix=${CMAKE_DIR} && \
+    rm ${CMAKE_SCRIPT}
+ENV PATH=${CMAKE_DIR}/bin:$PATH
+
+# Install Clang/LLVM
+ENV LLVM_DIR=/opt/llvm
+RUN LLVM_VERSION=7.0.1 && \
+    LLVM_URL=http://releases.llvm.org/${LLVM_VERSION}/clang+llvm-${LLVM_VERSION}-x86_64-linux-gnu-ubuntu-18.04.tar.xz && \
+    LLVM_ARCHIVE=llvm-${LLVM_VERSION}.tar.xz && \
```
[the remark this answers is not in the document]
Comment (dalg24) [reply docker/Dockerfile]: Now verifying the signature

File: docker/Dockerfile [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+
+# Install OpenMPI
+ENV OPENMPI_DIR=/opt/openmpi
+RUN OPENMPI_VERSION=3.1.3 && \
+    OPENMPI_VERSION_SHORT=3.1 && \
+    OPENMPI_SHA1=b3c60e2bdd5a8a8e758fd741f9a5bebb84da5e81 && \
+    OPENMPI_URL=https://download.open-mpi.org/release/open-mpi/v${OPENMPI_VERSION_SHORT}/openmpi-${OPENMPI_VERSION}.tar.bz2 && \
+    OPENMPI_ARCHIVE=openmpi-${OPENMPI_VERSION}.tar.bz2 && \
+    [ ! -z "${CUDA_VERSION}" ] && CUDA_OPTIONS=--with-cuda || true && \
+    export SCRATCH_DIR=${PWD}/scratch && mkdir -p ${SCRATCH_DIR} && cd ${SCRATCH_DIR} && \
```
[the remark this answers is not in the document]
Comment (dalg24) [reply docker/Dockerfile]: Fixed

**Comment by dalg24** [comment]
@aprokop I have stripped the docker-compose for the dev container
I will add it along with a configure script in another PR

**Pull request merged**